q2_moshpit/busco/busco.py

```diff
@@ -5,21 +5,104 @@
+
+from q2_moshpit.busco.plots_detailed import _draw_detailed_plots
+from q2_moshpit.busco.plots_summary import _draw_marker_summary_histograms, \
+    _draw_selectable_summary_histograms
+
 from q2_moshpit.busco.utils import (
-    _parse_busco_params,
-    _render_html,
+    _parse_busco_params, _collect_summaries, _rename_columns,
+    _parse_df_columns, _partition_dataframe, _calculate_summary_stats,
+    _get_feature_table, _cleanup_bootstrap, _get_mag_lengths
 )
-from q2_moshpit._utils import _process_common_input_params
+from q2_moshpit._utils import _process_common_input_params, run_command
 from q2_types.per_sample_sequences._format import MultiMAGSequencesDirFmt
 
 
-def evaluate_busco(
-    output_dir: str,
+def _run_busco(
+    output_dir: str, mags: MultiMAGSequencesDirFmt, params: List[str]
+) -> Dict[str, str]:
+    """Evaluates bins for all samples using BUSCO.
+
+    Args:
+        output_dir (str): Location where the final results should be stored.
+        mags (MultiMAGSequencesDirFmt): The mags to be analyzed.
+        params (List[str]): List of parsed arguments to pass to BUSCO.
+
+    Returns:
+        dict: Dictionary where keys are sample IDs and values are the paths
+            to the `batch_summary.txt` generated by BUSCO, e.g.
+            `tmp/busco_output/<sample_id>/batch_summary.txt`.
+    """
+    base_cmd = ["busco", *params]
+
+    manifest: pd.DataFrame = mags.manifest.view(pd.DataFrame)
+    manifest["sample_dir"] = manifest.filename.apply(
+        lambda x: os.path.dirname(x)
+    )
+
+    sample_dirs = manifest["sample_dir"].unique()
+
+    path_to_run_summaries = {}
+
+    # For every unique sample dir run busco
+    for sample_dir in sample_dirs:
+        sample = os.path.split(sample_dir)[-1]
+
+        cmd = deepcopy(base_cmd)
+        cmd.extend([
+            "--in",
+            sample_dir,
+            "--out_path",
+            output_dir,
+            "-o",
+            sample
+        ])
+        run_command(cmd)
+
+        path_to_run_summary = os.path.join(
+            output_dir, sample, "batch_summary.txt"
+        )
+        if os.path.isfile(path_to_run_summary):
+            path_to_run_summaries[sample] = path_to_run_summary
+        else:
+            raise FileNotFoundError(
+                f"BUSCO batch summary file {path_to_run_summary} not found."
+            )
+
+    return path_to_run_summaries
+
+
+def _busco_helper(bins, common_args):
+    with tempfile.TemporaryDirectory() as tmp:
+        path_to_run_summaries = _run_busco(
+            output_dir=os.path.join(tmp, "busco_output"),
+            mags=bins,
+            params=common_args,
+        )
+
+        all_summaries = _collect_summaries(
+            run_summaries_fp_map=path_to_run_summaries,
+        )
+    all_summaries = _rename_columns(all_summaries)
+
+    lengths = _get_mag_lengths(bins)
+    all_summaries = all_summaries.join(lengths, on="mag_id")
+
+    return all_summaries
+
+
+def _evaluate_busco(
```

Comment:
This comment is in tandem with the code suggestion in `evaluate_busco`.

- Delete all keyword arguments except for `bins`. 
- add a keyword argument named something like `common_args`. E.g. `common_args: List[str]`
- remove lines 127-135

Reply:
I cannot do that - this function is registered as an action hence it requires all the arguments spelled out.

Reply:
No, as I said, it is _registered_ as an action (see plugin_setup.py) and so it needs all the params defined here. Then, the pipeline fetches and runs this action, which means they have their params defined independently.

Reply:
And yes, technically, you can use it as a standalone action, although it is hidden from the user (as it starts with `_`).

Reply:
> No, as I said, it is _registered_ as an action (see plugin_setup.py) and so it needs all the params defined here. Then, the pipeline fetches and runs this action, which means they have their params defined independently.

Yes, I understand. But my point is that **you get to decide which parameters to define in the registration of `_evaluate_busco`**, whether it's the "same" parameters as in `evaluate_busco` or only one (my suggestion); a list of strings (the parsed parameters for the command). Ofc having this one-parameter setup does not make sense if one (user or developer) would like to use  `_evaluate_busco` as a standalone action.

Reply:
No, I do not really get to decide that. Both of these parameter sets need to be pretty much the same - I should not define the `_evaluate_busco` action with "common_args" as an allowed param as this would allow passing a list of pretty much any kinds of strings which defies the semantic validation principles. Also, this action is not exposed on the CLI but a user of the Python API can easily use it, just as I am using it in the pipeline. Hence, these parameters need to stay the same in both of those functions.